jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -2444,14 +2447,29 @@ public void retain()
         public boolean release()
         {
             data.release();
-            boolean result = counter.release();
-            if (result)
+
+            if (counter.release())
             {
                 notIdle();
                 stream.notIdle();
-                flowControl.onDataConsumed(HTTP2Session.this, stream, flowControlLength);
+
+                // once the reference count goes to zero, then it does not matter if the buffer is empty or not, as it is no
+                // longer available to the application to read, and thus has been consumed (if not already consumed).
+                if (consumed.compareAndSet(false, true))
+                    flowControl.onDataConsumed(HTTP2Session.this, stream, flowControlLength);
+                return true;
             }
-            return result;
+
+            // If all the data has been consumes and at least one release call is made, that indicates that the content
+            // has been fully consumed. The fact that release count has not gone to zero indicates that the application wants
+            // to retain the data, but not prevent more data being read.
+            // Note that a release with a non-empty buffer, may indicate the release of just a part of the buffer and thus
+            // the data has not been consumed. An optimization might be to consume the data up to the current position on each
+            // release.
+            if (BufferUtil.isEmpty(data.frame().getByteBuffer()) && consumed.compareAndSet(false, true))
```

Comment:
I'm not so sure about this one. How about the following oddity?

```java
Buffer duplicate = buffer.slice();
buffer.get(...); // read everything
duplicate.get(...); // try to read again
```

Wouldn't that be a nasty surprise?

Reply:
After discussing this issue with @sbordet, we came to an agreement that the right place to open the flow control window is in `HTTP2Stream.readData()`, i.e.: right before handing the buffer to the application code, to decouple the window control from the re-pooling of the buffer. This would make the behavior in line with how H1 works, as the TCP window is controlled by when you read from the socket.

Reply:
@lorban I don't understand the surprise?  If a make a duplicate, then of course I can read the data again from the buffer.

@sbordet @lorban Might gut reaction to the idea of opening the window in `readData` is NNNNNNNNOOOOOOOoooooo.... but I can't think of a rational reason why not.
That would make reading about consumption and retain/release about buffering.  Good separation of concerns.

Let me try to implement....